### PR TITLE
Tests stability

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Apart from https://github.com/Microsoft/RTVS/issues/2001 All tests pass consistently together on my two machines. Changes:

- Make sure editor tree and documents are disposed of. Otherwise they stay attached to idle and throw when called on wrong thread.
- Idle event can specify which thread to fire at.
- Some tests had hardcoded R 3.2, Now use RInstallation
- RInstallation static didn't work well since registry mocks affected tests that expected real registry when running in parallel
- Export provider should not be disposed since tests dispose it before WPF window is fully closed and all view and editor components are disposed of.

### CODE REVIEW

File: src/Common/Core/Impl/IO/FileSystem.cs [hunk 1]
```diff
@@ -8,6 +8,9 @@
 
 namespace Microsoft.Common.Core.IO {
     public sealed class FileSystem : IFileSystem {
+        private static FileSystem _instance = new FileSystem();
+        public static IFileSystem Current => _instance;
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/IO/FileSystem.cs]: Let's not introduce more singletons.
This class is lightweight and can be instantiated even when it isn't accessible from MEF

File: src/Host/Client/Test/ExecutionTracing/BreakpointsTest.cs [hunk 2]
```diff
@@ -18,6 +18,8 @@
 
 namespace Microsoft.R.ExecutionTracing.Test {
     [ExcludeFromCodeCoverage]
+    [Category.R.ExecutionTracing]
+    [Collection(CollectionNames.NonParallel)]
```
Comment (AlexanderSher) [inline src/Host/Client/Test/ExecutionTracing/BreakpointsTest.cs]: It is important for these tests to run in parallel - that simulates multiple instances of RHost running concurrently

File: src/R/Components/Test/History/RHistoryIntegrationTest.cs [hunk 3]
```diff
@@ -158,9 +158,5 @@ public class RHistoryIntegrationTest : IDisposable {
                 text.Should().Be(string.Empty);
             }
         }
-
-        public void Dispose() {
-            (_exportProvider as IDisposable)?.Dispose();
```
Comment (AlexanderSher) [inline src/R/Components/Test/History/RHistoryIntegrationTest.cs]: What is this change for?

File: src/R/Components/Test/MefCompositionTests.cs [hunk 4]
```diff
@@ -56,9 +56,5 @@ public class MefCompositionTests : IDisposable {
         public void InteractiveWindowFactoryService() {
             _exportProvider.GetExportedValue<IInteractiveWindowFactoryService>().Should().NotBeNull();
         }
-
-        public void Dispose() {
-            (_exportProvider as IDisposable)?.Dispose();
```
Comment (AlexanderSher) [inline src/R/Components/Test/MefCompositionTests.cs]: Disposing `ExportProvider` helps detecting static caches of MEF components, which should be considered an anti-pattern (MEF can be forced to rebuild its instances, while static cache will have no idea about it).

File: src/Common/Core/Impl/IO/FileSystem.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/IO/FileSystem.cs]: OK. It was for convenience only, but yeah, we can just new it.

File: src/R/Components/Test/MefCompositionTests.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/R/Components/Test/MefCompositionTests.cs]: We need to figure out then when to dispose it. disposing it here causes core editor to throw since although from our point of view the control is closed, it is still getting some code executed (I guess dispatcher frames age getting pushed). I'll see if I can find safe place to disposed of the container.

File: src/R/Components/Test/History/RHistoryIntegrationTest.cs (hunk 3 above)
Comment (MikhailArkhipov) [inline src/R/Components/Test/History/RHistoryIntegrationTest.cs]: See the other comment wrt export provider disposal

File: src/Host/Client/Test/ExecutionTracing/BreakpointsTest.cs (hunk 2 above)
Comment (MikhailArkhipov) [inline src/Host/Client/Test/ExecutionTracing/BreakpointsTest.cs]: OK, it was attempt to make debugger breakpoints setting tests pass but this they still fail, i can put it back.

File: src/R/Components/Test/MefCompositionTests.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/R/Components/Test/MefCompositionTests.cs]: OK, restored. couple throw in debug, but it doesnt seem to affect the result

File: src/Common/Core/Impl/Shell/IIdleTimeService.cs [hunk 5]
```diff
@@ -2,8 +2,10 @@
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
 
+using System.Threading;
+
 namespace Microsoft.Languages.Editor.Host {
     public interface IIdleTimeService {
-        void DoIdle();
+        void DoIdle(Thread thread = null);
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Shell/IIdleTimeService.cs]: Can you explain this change?

Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Shell/IIdleTimeService.cs]: When tests run in parallel while EditorShell is a singleton, they may get events fired from a wrong thread. In most cases it doesn't matter though, but I've seen some fail occasionally. However, with changes where editor objects are disposed properly, this may not be that necessary.

File: src/R/Components/Impl/History/Implementation/RHistoryProvider.cs [hunk 6]
```diff
@@ -16,18 +16,29 @@ namespace Microsoft.R.Components.History.Implementation {
...
-        private readonly IFileSystem _fileSystem;
         private readonly IEditorOperationsFactoryService _editorOperationsFactory;
         private readonly IRtfBuilderService _rtfBuilderService;
         private readonly ITextSearchService2 _textSearchService;
         private readonly IRSettings _settings;
         private readonly IContentTypeRegistryService _contentTypeRegistryService;
         private readonly Dictionary<ITextBuffer, IRHistory> _histories;
 
+        private IFileSystem _fileSystem;
+        internal IFileSystem FileSystem {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: Where this change is used?

File: src/Languages/Editor/Impl/Shell/IEditorShell.cs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
+using System.Threading;
```
Comment (AlexanderSher) [inline src/Languages/Editor/Impl/Shell/IEditorShell.cs]: You can undo this change

File: src/Host/Client/Test/ExecutionTracing/SteppingTest.cs [hunk 8]
```diff
@@ -20,6 +19,8 @@
 
 namespace Microsoft.R.ExecutionTracing.Test {
     [ExcludeFromCodeCoverage]
+    [Category.R.ExecutionTracing]
+    [Collection(CollectionNames.NonParallel)]
```
Comment (AlexanderSher) [inline src/Host/Client/Test/ExecutionTracing/SteppingTest.cs]: Do you need NonParallel here?

File: src/Host/Client/Impl/Definitions/PlotMessage.cs [hunk 9]
```diff
@@ -17,11 +17,11 @@ public struct PlotMessage {
 
         public bool IsClearAll => string.IsNullOrEmpty(FilePath);
 
-        public bool IsPlot => new FileInfo(FilePath).Length > 0;
+        public bool IsPlot => File.Exists(FilePath) && new FileInfo(FilePath).Length > 0;
```
Comment (AlexanderSher) [inline src/Host/Client/Impl/Definitions/PlotMessage.cs]: Technically, file can exists during first check and be deleted during second change. We need to wrap `new FileInfo(FilePath).Length > 0` into try/catch

File: src/R/Components/Impl/History/Implementation/RHistoryProvider.cs (hunk 6 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: Added so substitute could be provided as in other places, but if you think it won't be needed we can nix it

Comment (AlexanderSher) [inline src/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: Is it required in any tests right now?

---
In reply to: [68613538](https://github.com/Microsoft/RTVS/pull/2004#discussion_r68613538) [](ancestors = 68613538)

Comment (AlexanderSher) [inline src/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: AFAIK, in RHistory tests, IFileSystem implementation is provided through MEF, so you can leave it in ctor but mark parameter with AllowDefault

---
In reply to: [68614367](https://github.com/Microsoft/RTVS/pull/2004#discussion_r68614367) [](ancestors = 68614367,68613538)

Comment (MikhailArkhipov) [inline src/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: IFileSystem is no longer exported

File: src/Host/Client/Test/ExecutionTracing/SteppingTest.cs (hunk 8 above)
Comment (MikhailArkhipov) [inline src/Host/Client/Test/ExecutionTracing/SteppingTest.cs]: BreakContinue is failing unless I put it here.

Comment (AlexanderSher) [inline src/Host/Client/Test/ExecutionTracing/SteppingTest.cs]: Can you please create a bug and assign it to @int19h ? This shouldn't happen.

File: src/R/Components/Impl/History/Implementation/RHistoryProvider.cs (hunk 6 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: In R.Components.Tests it is explicitly added to the container:
```{csharp}
        protected override void AddValues(CompositionContainer container) {
            base.AddValues(container);
            var coreShell = new TestCoreShell(container);
            var batch = new CompositionBatch()
                .AddValue(FileSystemStubFactory.CreateDefault())
                .AddValue<IRSettings>(RSettingsStubFactory.CreateForExistingRPath())
                .AddValue<ICoreShell>(coreShell)
                .AddValue(coreShell);
            container.Compose(batch);
        }
```

Comment (MikhailArkhipov) [inline src/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: I guess it should be removed from the container then. Frankly I'd rather have containers use actual exported values rather than manually add stuff since it makes container != runtime container.

Comment (AlexanderSher) [inline src/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: They aren't identical. Runtime container provides more exports and may provide them in a different time periods.
Injecting `IFileSystem` in constructor simplifies test code. For example, `RInteractiveWorkflow` calls `CreateRHistory` right in constructor, so we have to explicitly call `_exportProvider.GetExportedValue<IRHistoryProvider>().FileSystem = FileSystemStubFactory.CreateDefault()` in test constructors.
I don't see big difference in default value checking inside constructor or in the internal property. In both cases we know that it is null in production and not null in unit tests.

**Comment by AlexanderSher** [comment]
:+1:
